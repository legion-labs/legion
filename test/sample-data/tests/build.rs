use std::{
    env,
    path::{Path, PathBuf},
};

use legion_data_compiler::Manifest;

pub fn target_dir() -> PathBuf {
    env::current_exe()
        .ok()
        .map(|mut path| {
            path.pop();
            if path.ends_with("deps") {
                path.pop();
            }
            path
        })
        .unwrap_or_else(|| panic!("cannot find test directory"))
}

pub fn data_build_exe() -> PathBuf {
    target_dir().join(format!("data-build{}", env::consts::EXE_SUFFIX))
}

fn exec_create_build_index(
    path: impl AsRef<Path>,
    project: impl AsRef<Path>,
) -> std::io::Result<std::process::Output> {
    let mut command = std::process::Command::new(data_build_exe());
    command.arg("create");
    command.arg(path.as_ref().to_str().unwrap());
    command.arg(format!("--project={}", project.as_ref().to_str().unwrap()));
    let output = command.output()?;
    assert!(output.status.success());
    Ok(output)
}

fn exec_data_compile(
    compile_path: &str,
    buildindex_path: &Path,
    destination: &Path,
) -> std::io::Result<std::process::Output> {
    let target = "game";
    let platform = "windows";
    let locale = "en";
    let mut command = std::process::Command::new(data_build_exe());
    command.arg("compile");
    command.arg(compile_path);
    command.arg(format!("--cas={}", destination.to_str().unwrap()));
    command.arg(format!("--target={}", target));
    command.arg(format!("--platform={}", platform));
    command.arg(format!("--locale={}", locale));
    command.arg(format!(
        "--buildindex={}",
        buildindex_path.to_str().unwrap()
    ));
    let output = command.output()?;
    assert!(output.status.success());
    Ok(output)
}

/// read the build index and remove non-deterministic data.
fn read_build_index(buildindex_path: &Path) -> String {
    let content = std::fs::read_to_string(&buildindex_path).expect("file content");
    let mut content = json::parse(&content).expect("valid json");
    content.remove("project_index");
    content.pretty(2)
}

#[test]
fn incremental_build() {
    let work_dir = tempfile::tempdir().unwrap();
    let sampledata_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let project = sampledata_dir.join("project.index");
    let buildindex = work_dir.as_ref().join("build.index");

    // create build index and do a source pull
    exec_create_build_index(&buildindex, project).expect("new build index");

    insta::assert_snapshot!("initial_index", read_build_index(&buildindex));

    // default root object in sample data
    // /world/sample_1.ent (offline_entity) => runtime_entity
    let root_entity = "d004cd1c00000000fcd3242ec9691beb|019c8223";

    //
    // first data build
    //
    let out =
        exec_data_compile(root_entity, &buildindex, work_dir.path()).expect("build completed");

    let manifest: Manifest = serde_json::from_slice(&out.stdout).expect("valid manifest");
    insta::assert_json_snapshot!("first_manifest", manifest);

    let first_buildindex = read_build_index(&buildindex);
    insta::assert_snapshot!("first_index", first_buildindex);

    //
    // incremental data build
    //
    let out =
        exec_data_compile(root_entity, &buildindex, work_dir.path()).expect("build completed");

    let incremental_manifest: Manifest =
        serde_json::from_slice(&out.stdout).expect("valid manifest");

    insta::assert_json_snapshot!("incremental_manifest", incremental_manifest);

    let incremental_buildindex = read_build_index(&buildindex);
    insta::assert_snapshot!("incremental_index", incremental_buildindex);

    // incremental build with no changes should not affect the build index
    assert_eq!(first_buildindex, incremental_buildindex);
}
