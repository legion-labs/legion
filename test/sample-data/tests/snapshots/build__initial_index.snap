---
source: test/sample-data/tests/build.rs
expression: empty_buildindex

---
{
  "version": "0.1.0",
  "resources": [
    {
      "id": "13b5a84e0000000061f6de1be4764697",
      "dependencies": [],
      "resource_hash": "f480b17eea122496"
    },
    {
      "id": "13b5a84e0000000061f6de1be4764697|74dc0e53_albedo|f9c9670d",
      "dependencies": [
        "13b5a84e0000000061f6de1be4764697|74dc0e53_albedo"
      ],
      "resource_hash": null
    },
    {
      "id": "13b5a84e0000000061f6de1be4764697|74dc0e53_metalness|f9c9670d",
      "dependencies": [
        "13b5a84e0000000061f6de1be4764697|74dc0e53_metalness"
      ],
      "resource_hash": null
    },
    {
      "id": "13b5a84e0000000061f6de1be4764697|74dc0e53_normal|f9c9670d",
      "dependencies": [
        "13b5a84e0000000061f6de1be4764697|74dc0e53_normal"
      ],
      "resource_hash": null
    },
    {
      "id": "13b5a84e0000000061f6de1be4764697|74dc0e53_roughness|f9c9670d",
      "dependencies": [
        "13b5a84e0000000061f6de1be4764697|74dc0e53_roughness"
      ],
      "resource_hash": null
    },
    {
      "id": "2b368fed00000000f8b1bd1dceaf6f33",
      "dependencies": [
        "13b5a84e0000000061f6de1be4764697|74dc0e53_albedo|f9c9670d",
        "13b5a84e0000000061f6de1be4764697|74dc0e53_metalness|f9c9670d",
        "13b5a84e0000000061f6de1be4764697|74dc0e53_normal|f9c9670d",
        "13b5a84e0000000061f6de1be4764697|74dc0e53_roughness|f9c9670d"
      ],
      "resource_hash": "b69f41dbe609db99"
    },
    {
      "id": "2b368fed00000000f8b1bd1dceaf6f33|669eb7d0",
      "dependencies": [
        "2b368fed00000000f8b1bd1dceaf6f33"
      ],
      "resource_hash": null
    },
    {
      "id": "417862e5000000005127bdca42145845",
      "dependencies": [
        "d004cd1c0000000062aa07fb54e7a8ee|019c8223"
      ],
      "resource_hash": "e7e15e0065e6b19d"
    },
    {
      "id": "417862e5000000005127bdca42145845|0cd05ad8",
      "dependencies": [
        "417862e5000000005127bdca42145845"
      ],
      "resource_hash": null
    },
    {
      "id": "417862e5000000007a80bfadcb6549e7",
      "dependencies": [
        "d004cd1c0000000091d021d6f4e05efe|019c8223"
      ],
      "resource_hash": "8d553443b452074b"
    },
    {
      "id": "417862e5000000007a80bfadcb6549e7|0cd05ad8",
      "dependencies": [
        "417862e5000000007a80bfadcb6549e7"
      ],
      "resource_hash": null
    },
    {
      "id": "417862e5000000007df5d0fcdb1eb69b",
      "dependencies": [
        "d004cd1c0000000091d021d6f4e05efe|019c8223"
      ],
      "resource_hash": "8d553443b452074b"
    },
    {
      "id": "417862e5000000007df5d0fcdb1eb69b|0cd05ad8",
      "dependencies": [
        "417862e5000000007df5d0fcdb1eb69b"
      ],
      "resource_hash": null
    },
    {
      "id": "417862e50000000090b1c112655b00e8",
      "dependencies": [
        "d004cd1c0000000091d021d6f4e05efe|019c8223"
      ],
      "resource_hash": "8d553443b452074b"
    },
    {
      "id": "417862e50000000090b1c112655b00e8|0cd05ad8",
      "dependencies": [
        "417862e50000000090b1c112655b00e8"
      ],
      "resource_hash": null
    },
    {
      "id": "417862e500000000ce1185cf2b953536",
      "dependencies": [
        "d004cd1c0000000091d021d6f4e05efe|019c8223"
      ],
      "resource_hash": "8d553443b452074b"
    },
    {
      "id": "417862e500000000ce1185cf2b953536|0cd05ad8",
      "dependencies": [
        "417862e500000000ce1185cf2b953536"
      ],
      "resource_hash": null
    },
    {
      "id": "417862e500000000e578fadce0648bcc",
      "dependencies": [
        "d004cd1c0000000091d021d6f4e05efe|019c8223"
      ],
      "resource_hash": "8d553443b452074b"
    },
    {
      "id": "417862e500000000e578fadce0648bcc|0cd05ad8",
      "dependencies": [
        "417862e500000000e578fadce0648bcc"
      ],
      "resource_hash": null
    },
    {
      "id": "4e1dd441000000007714d9ad404557f7",
      "dependencies": [
        "2b368fed00000000f8b1bd1dceaf6f33|669eb7d0"
      ],
      "resource_hash": "6c8072661bdeec94"
    },
    {
      "id": "4e1dd441000000007714d9ad404557f7|ea2510e8",
      "dependencies": [
        "4e1dd441000000007714d9ad404557f7"
      ],
      "resource_hash": null
    },
    {
      "id": "9f639fe800000000d4b05b1a98368098",
      "dependencies": [],
      "resource_hash": "4710a4f11115835a"
    },
    {
      "id": "d004cd1c00000000519d4a9f47df0ce5",
      "dependencies": [
        "4e1dd441000000007714d9ad404557f7|ea2510e8",
        "4e1dd441000000007714d9ad404557f7|ea2510e8"
      ],
      "resource_hash": "389e5bf518b9c948"
    },
    {
      "id": "d004cd1c00000000519d4a9f47df0ce5|019c8223",
      "dependencies": [
        "d004cd1c00000000519d4a9f47df0ce5"
      ],
      "resource_hash": null
    },
    {
      "id": "d004cd1c0000000062aa07fb54e7a8ee",
      "dependencies": [
        "417862e500000000ce1185cf2b953536|0cd05ad8",
        "417862e500000000e578fadce0648bcc|0cd05ad8"
      ],
      "resource_hash": "22204ddc3bbb42a5"
    },
    {
      "id": "d004cd1c0000000062aa07fb54e7a8ee|019c8223",
      "dependencies": [
        "d004cd1c0000000062aa07fb54e7a8ee"
      ],
      "resource_hash": null
    },
    {
      "id": "d004cd1c0000000091d021d6f4e05efe",
      "dependencies": [],
      "resource_hash": "a7a94184f11d5ade"
    },
    {
      "id": "d004cd1c0000000091d021d6f4e05efe|019c8223",
      "dependencies": [
        "d004cd1c0000000091d021d6f4e05efe"
      ],
      "resource_hash": null
    },
    {
      "id": "d004cd1c00000000fcd3242ec9691beb",
      "dependencies": [
        "417862e5000000005127bdca42145845|0cd05ad8",
        "417862e5000000007a80bfadcb6549e7|0cd05ad8",
        "417862e5000000007df5d0fcdb1eb69b|0cd05ad8",
        "417862e50000000090b1c112655b00e8|0cd05ad8",
        "d004cd1c00000000519d4a9f47df0ce5|019c8223"
      ],
      "resource_hash": "2c6ddab0baf21b24"
    }
  ],
  "compiled_resources": [],
  "compiled_resource_references": [],
  "pathid_mapping": {
    "019c822300000000264b29e1daf98c75": "d004cd1c0000000091d021d6f4e05efe|019c8223",
    "019c82230000000052d85baf2a2c79a4": "d004cd1c0000000062aa07fb54e7a8ee|019c8223",
    "019c822300000000b4e464b9fef217e6": "d004cd1c00000000519d4a9f47df0ce5|019c8223",
    "0cd05ad8000000005aa0bf3c69af020f": "417862e500000000ce1185cf2b953536|0cd05ad8",
    "0cd05ad800000000691906ec63d6d8d8": "417862e5000000007df5d0fcdb1eb69b|0cd05ad8",
    "0cd05ad8000000009ce518e3cf3927a3": "417862e500000000e578fadce0648bcc|0cd05ad8",
    "0cd05ad800000000a5c01d33d46b3d82": "417862e50000000090b1c112655b00e8|0cd05ad8",
    "0cd05ad800000000bfb87946adf03803": "417862e5000000005127bdca42145845|0cd05ad8",
    "0cd05ad800000000f271e52bf0a8c4f1": "417862e5000000007a80bfadcb6549e7|0cd05ad8",
    "13b5a84e0000000061f6de1be4764697": "13b5a84e0000000061f6de1be4764697",
    "2b368fed00000000f8b1bd1dceaf6f33": "2b368fed00000000f8b1bd1dceaf6f33",
    "417862e5000000005127bdca42145845": "417862e5000000005127bdca42145845",
    "417862e5000000007a80bfadcb6549e7": "417862e5000000007a80bfadcb6549e7",
    "417862e5000000007df5d0fcdb1eb69b": "417862e5000000007df5d0fcdb1eb69b",
    "417862e50000000090b1c112655b00e8": "417862e50000000090b1c112655b00e8",
    "417862e500000000ce1185cf2b953536": "417862e500000000ce1185cf2b953536",
    "417862e500000000e578fadce0648bcc": "417862e500000000e578fadce0648bcc",
    "4e1dd441000000007714d9ad404557f7": "4e1dd441000000007714d9ad404557f7",
    "669eb7d000000000c9875c015d617949": "2b368fed00000000f8b1bd1dceaf6f33|669eb7d0",
    "74dc0e530000000011ba23150911dc55": "13b5a84e0000000061f6de1be4764697|74dc0e53_albedo",
    "74dc0e53000000006299a94ba17af71b": "13b5a84e0000000061f6de1be4764697|74dc0e53_normal",
    "74dc0e530000000067c73bfed25d45cf": "13b5a84e0000000061f6de1be4764697|74dc0e53_metalness",
    "74dc0e53000000009df90f0734078b5f": "13b5a84e0000000061f6de1be4764697|74dc0e53_roughness",
    "9f639fe800000000d4b05b1a98368098": "9f639fe800000000d4b05b1a98368098",
    "d004cd1c00000000519d4a9f47df0ce5": "d004cd1c00000000519d4a9f47df0ce5",
    "d004cd1c0000000062aa07fb54e7a8ee": "d004cd1c0000000062aa07fb54e7a8ee",
    "d004cd1c0000000091d021d6f4e05efe": "d004cd1c0000000091d021d6f4e05efe",
    "d004cd1c00000000fcd3242ec9691beb": "d004cd1c00000000fcd3242ec9691beb",
    "ea2510e8000000007354d8806dd36e41": "4e1dd441000000007714d9ad404557f7|ea2510e8",
    "f9c9670d0000000062572ca04da3b129": "13b5a84e0000000061f6de1be4764697|74dc0e53_roughness|f9c9670d",
    "f9c9670d00000000b2333bed9c99dec1": "13b5a84e0000000061f6de1be4764697|74dc0e53_albedo|f9c9670d",
    "f9c9670d00000000e9672e9a29dbae26": "13b5a84e0000000061f6de1be4764697|74dc0e53_normal|f9c9670d",
    "f9c9670d00000000f343b5e9436e3bd5": "13b5a84e0000000061f6de1be4764697|74dc0e53_metalness|f9c9670d"
  }
}
